File: dss/linked_list/py/linked_list.py
from typing import Generator


class Node:
    def __init__(self, key: int) -> None:
        self.key = key
        self.next: Node | None = None
    

class LinkedList:
    def __init__(self) -> None:
        self.head: Node | None = None

    def __iter__(self) -> Generator[Node, None, None]:
        node = self.head
        while node:
            yield node
            node = node.next
    
    def __str__(self) -> str:
        return "->".join([f"{node.key}" for node in self])

    def __len__(self) -> int:
        counter = 0
        for _ in self:
            counter += 1
        return counter
    
    def build(self, keys: list[int]) -> None:
        for key in keys:
            self.insert_last(key)

    def insert_first(self, key: int) -> None:
        new_node = Node(key)
        if self.head == None:
            self.head = new_node
            return
        new_node.next = self.head
        self.head = new_node
            
    def insert_last(self, key: int) -> None:
        new_node = Node(key)
        skip = 0
        if self.head == None:
            self.head = new_node
            skip += 1
        
        current_node = self.head
        for _ in range(len(self) - skip):
            if current_node.next == None:
                current_node.next = Node(key)
            current_node = current_node.next

    def insert_at(self, index: int, key: int) -> None:
        raise NotImplementedError

    def get_at(self, i: int) -> None:
        raise NotImplementedError

    def set_at(self, i: int, key: int) -> None:
        raise NotImplementedError

    def get_by_key(self, key: int) -> None:
        raise NotImplementedError

    def delete_by_key(self, key: int) -> None:
        raise NotImplementedError
    
    def reverse(self) -> None:
        raise NotImplementedError

File: dss/linked_list/c/linked_list.c
#include<stdio.h>
#include<stdlib.h>

typedef struct node {
    int key;
    struct node *next;
} node;

node* new_linked_list(const int *keys, size_t size) {
    node *list = NULL;
    node *last_node = NULL;
    for (int i = 0; i < size; i++) {
        node *new_node = malloc(sizeof(node));
        new_node->key = keys[i];
        new_node->next = NULL;
        if (last_node != NULL) {
            last_node->next = new_node;
        } else {
            list = new_node;
        }
        last_node = new_node;
    }
    return list;
}

// node* insert_at(node *list, int index, int key) {
//     return NULL;
// }

node* insert_first(node *list, int key) {
    if (list == NULL) {
        int keys[1] = {key};
        list = new_linked_list(keys, (size_t)1);
        return list;
    }
    node *new_node = malloc(sizeof(node));
    new_node->key = key;
    new_node->next=list;

    return new_node;
}

node* insert_last(node *list, int key) {
    if (list == NULL) {
        return NULL;
    }
    node *current_node = list;
    while (current_node->next != NULL) {
        current_node = current_node->next;
    }
    node *new_node = malloc(sizeof(node));
    new_node->key = key;
    new_node->next = NULL;
    current_node->next = new_node;
    return new_node;
}

node* get_node(node *list, int key) {
    node *current_node = list;
    while (current_node->next != NULL && current_node->key != key) {
        current_node = current_node->next;
    }
    if (current_node->key == key) {
        return current_node;
    }
    return NULL;
}

void delete_node(node *list, int key) {
    node *previous_node = NULL, *current_node = list;

    while (current_node != NULL && current_node->key != key) {
        previous_node = current_node;
        current_node = current_node->next;
    }
    if (current_node == NULL) {
        return;
    }
    previous_node->next = current_node->next;
    free(current_node);
}

void print_keys(node *list) {
    node *ptr = list;
    while (ptr != NULL) {
        printf("%d->", ptr->key);
        ptr = ptr->next;
    }
}

void free_list(node *list) {
    node *ptr = list;
    while (ptr != NULL) {
        node *next = ptr->next;
        free(ptr);
        ptr = next;
    }
}

int main(void) {
    int keys[5] = {1, 2, 3, 4, 5};
    node *list = NULL;
    list = insert_first(list, 100);
    print_keys(list);
    free_list(list);
}